server/src/main/java/org/opensearch/index/engine/Engine.java

```diff
@@ -1291,6 +1300,14 @@ public void failEngine(String reason, @Nullable Exception failure) {
                             );
                         }
                     }
+
+                    // If cleanup unreferenced flag is enabled and force merge or regular merge failed due to disk full,
+                    // cleanup all unreferenced files created during failed and reset the shard state back to last
+                    // Lucene Commit.
+                    if (shouldCleanupUnreferencedFiles() && isMergeFailureDueToDiskFull(failure, reason)) {
+                        cleanUpUnreferencedFiles();
+                    }
```

Comment:
Can we use `onMergeFinished` event?

Reply:
I think you meant ```handleMergeException``` provided by EngineMergeScheduler as merge has not finished but failed. This seems similar to approach of failing IndexWriter discussed in this [RFC](https://github.com/opensearch-project/OpenSearch/issues/8024#issuecomment-1592429149). As discussed, this can cause a lot of issues:

1. This approach is not a clean approach as we need to handle multiple edge case scenarios like prevent reader refresh when clean up is ongoing.
2. Callback handleMergeException provided by EngineMergeScheduler is called whenever any segment merge fails. If multiple merges are running at the same time for the same Shard (but different segments) we need to handle multithreading scenarios like cleanup should happen only from one thread.
3. Since shard has not failed yet, it can be possible that another parallel operation like relocation, etc gets triggered when unreferenced file clean up is ongoing. This may corrupt the shard.